### PR TITLE
Feat/multi vault note creation

### PR DESCRIPTION
**Pull request by jonathanyeung**
Related to https://github.com/dendronhq/dendron/issues/649

On Note Creation in a multi-vault environment, this change will try to pick the correct vault location for the new note based on two factors - hierarchy match and note context.

### CODE REVIEW

File: packages/plugin-core/src/test/suite-integ/LookupUtils.test.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        genRandomId: true,
+      });
+    });
+  }
+}
+
+/**
+ * Tests the Vault Recommendation For New Notes Functionality in Utils
+ */
+suite.only("Lookup Utils Test", function runSuite() {
```
Comment (kevinslin) [inline packages/plugin-core/src/test/suite-integ/LookupUtils.test.ts:69]: remove `.only`. also, feel free to add it to the [husky](https://github.com/dendronhq/dendron/blob/dev/hooks/pre-push.js#L20:L20) scripit

File: packages/plugin-core/src/components/lookup/utils.ts [hunk 2]
```diff
@@ -390,16 +407,180 @@ export class PickerUtilsV2 {
...
-    >;
-    return resp;
+    const resp = await VSCodeUtils.showQuickPick(items, {
+      title: "Select Vault",
+    });
+
+    return resp ? resp.vault : undefined;
+  }
+
+  // Determine which are the most appropriate vault(s) to create this note in.
```
Comment (kevinslin) [inline packages/plugin-core/src/components/lookup/utils.ts:462]: use 
```
/**
 *
 **/
```

format comments so they show up in docstrings

File: packages/plugin-core/src/components/lookup/utils.ts [hunk 3]
```diff
@@ -390,16 +407,180 @@ export class PickerUtilsV2 {
...
+    let vaultSuggestions: VaultPickerItem[] = [];
+
+    const engine = getWS().getEngine();
+
+    // Only 1 vault, no other options to choose from:
+    if (engine.vaults.length <= 1) {
+      return Array.of({ vault });
+    }
+
+    const hierarchies = fname.split(".").slice(0, -1);
```
Comment (kevinslin) [inline packages/plugin-core/src/components/lookup/utils.ts:483]: we refer to the top part of a hierarchy as the `domain`. there's actually a utility method to get it [here](https://github.com/dendronhq/dendron/blob/feat/multi-vault-note-creation/packages/common-all/src/dnode.ts#L243:L243). its not used anywhere so you are free to use it.

it currently throws an error if the node you pass in is the root node but feel free to change it to return the root instead

File: packages/plugin-core/src/test/suite-integ/LookupUtils.test.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  const ctx: vscode.ExtensionContext = setupBeforeAfter(this, {});
+  this.timeout(TIMEOUT);
```
Comment (kevinslin) [inline packages/plugin-core/src/test/suite-integ/LookupUtils.test.ts:71]: this is already done in `setupBeforeAfter` so you can remove

**Review by kevinslin — CHANGES_REQUESTED** [review]

File: packages/plugin-core/src/test/suite-integ/LookupUtils.test.ts (hunk 1 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/test/suite-integ/LookupUtils.test.ts:69]: good catch :) let me add

File: packages/plugin-core/src/components/lookup/utils.ts (hunk 3 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/components/lookup/utils.ts:483]: I see - I've diverged a bit here though; whereas the `DNodeUtils` uses `DNodeProps` and `DNodePropsDict`, I stripped down the args on this method to just have fname of the proposed new note.

File: packages/plugin-core/src/test/suite-integ/LookupUtils.test.ts (hunk 1 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/test/suite-integ/LookupUtils.test.ts:69]: will add husky change as part of a separate review.

**Review by kevinslin — CHANGES_REQUESTED** [review]
looks like tests are failing due to unused variable

**Comment by kevinslin** [comment]
Looks like tests are now failing on scratch notes

**Comment by jonathanyeung** [comment]
> Looks like tests are now failing on scratch notes

Apologies haha - I actually have some issues running tests locally as I hit some transient issues, so I end up relying on the Github test pass...dunno if it's a Windows problem.  Let me try to dig in

**Comment by kevinslin** [comment]
looks like tests are still failing

```
 @dendronhq/plugin-core:   1) Lookup, notesv2
@dendronhq/plugin-core:        onAccept with modifiers
@dendronhq/plugin-core:          with lookupPrompt on other vault:
@dendronhq/plugin-core:      Error: Timeout of 300000ms exceeded. For
```

**Comment by jonathanyeung** [comment]
> looks like tests are still failing
> 
> ```
>  @dendronhq/plugin-core:   1) Lookup, notesv2
> @dendronhq/plugin-core:        onAccept with modifiers
> @dendronhq/plugin-core:          with lookupPrompt on other vault:
> @dendronhq/plugin-core:      Error: Timeout of 300000ms exceeded. For
> ```

This spiraled into me finally getting around to this: https://github.com/dendronhq/dendron/pull/980

Basically it was impossible for me to cleanly run tests locally haha.  I might just rebase on 980 so that I can see what other tests are getting affected by these changes.

**Review by kevinslin — APPROVED** [review]

**Pull request merged**